eth/signer.go

```diff
@@ -132,24 +115,123 @@ func (s *Signer) CompactSign(buf []byte) ([]byte, error) {
+	return append(buf, nonce[:]...), nonce, nil
+}
+
+// ToEthSignedMessageHash converts a given message to conform to the signed data
+// standard according to EIP-191.
+func ToEthPersonalSignedMessage(message []byte) []byte {
+	prefix := []byte(fmt.Sprintf("\x19Ethereum Signed Message:\n%d", len(message)))
+	return append(prefix, message...)
+}
+
+type signOpts struct {
+	raw, compact, personal, withNonce bool
+}
+
+// sign signs a given buffer depending on the chosen options:
+// withNonce = true, appends a nonce to the message
+// compact = true, returns a compactified version of the signature according to
+// EIP-2098.
+// personal = true, adds a prefix to the message to conform to the EIP-191
+// personal message standard.
+// raw = false, the message is hashed before signing
+func (s *Signer) sign(buf []byte, opts signOpts) ([]byte, *[32]byte, error) {
+	var nonce *[32]byte
+	var err error
+
+	if opts.withNonce {
+		var n [32]byte
+		buf, n, err = AppendRandomNonce(buf)
+		nonce = &n
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	if opts.personal {
+		buf = ToEthPersonalSignedMessage(buf)
+	}
+
+	if !opts.raw {
+		buf = crypto.Keccak256(buf)
+	}
+
+	sig, err := crypto.Sign(buf, s.key)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if !opts.compact {
+		return sig, nonce, nil
+	}
+
+	sig, err = CompactSignature(sig)
 	if err != nil {
-		return nil, nonce, err
+		return nil, nil, err
 	}
+
 	return sig, nonce, nil
 }
 
-// SignAddress is a convenience wrapper for s.CompactSign(addr.Bytes()).
-func (s *Signer) SignAddress(addr common.Address) ([]byte, error) {
-	return s.CompactSign(addr.Bytes())
+// RawSign returns an ECDSA signature of buf. USE WITH CAUTION as signed data
+// SHOULD be hashed first to avoid chosen-plaintext attacks. Prefer
+// Signer.Sign().
+func (s *Signer) RawSign(buf []byte) ([]byte, error) {
+	sig, _, err := s.sign(buf, signOpts{
+		raw:       true,
+		compact:   false,
+		personal:  false,
+		withNonce: false,
+	})
+	return sig, err
+}
+
+// Sign returns an ECDSA signature of keccak256(buf).
+func (s *Signer) Sign(buf []byte) ([]byte, error) {
+	sig, _, err := s.sign(buf, signOpts{
+		raw:       false,
+		compact:   false,
+		personal:  false,
+		withNonce: false,
+	})
+	return sig, err
+}
+
+// PersonalSign returns an EIP-191 conform personal ECDSA signature of buf
+// Convenience wrapper for s.CompactSign(toEthPersonalSignedMessage(buf))
+func (s *Signer) PersonalSign(buf []byte) ([]byte, error) {
+	sig, _, err := s.sign(buf, signOpts{
+		raw:       false,
+		compact:   true,
+		personal:  true,
+		withNonce: false,
+	})
+	return sig, err
+}
+
+// PersonalSignWithNonce generates a 32-byte nonce with crypto/rand and returns
+// s.PersonalSign(append(buf, nonce)).
+func (s *Signer) PersonalSignWithNonce(buf []byte) ([]byte, [32]byte, error) {
+	sig, nonce, err := s.sign(buf, signOpts{
+		raw:       false,
+		compact:   true,
+		personal:  true,
+		withNonce: true,
+	})
+	return sig, *nonce, err
```

Comment:
This will panic if there's an error returned by `sign()` and hence the `nonce` pointer is nil. First check for the error.

The fact that we're struggling with this is a definite code smell. Submit as is and I'll think it through a bit more. The problem comes from using a `[32]byte` as it's not nillable, which is fairly atypical in this style of Go. Perhaps we can move to the [uint256](https://pkg.go.dev/github.com/holiman/uint256) package?

Reply:
Fixed it. Yes, since we will mostly deal with `uint256` anyways it might make sense to rely on this package instead of `math/big` for added type safety. Is there maybe another package that also defines other types such as `uint128` etc.?

Reply:
Not that I know of. I suggested that specific uint256 because it's used within go-ethereum.